File: docs/DSA/src/py/subtree_of_another_tree.py
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def z_function(self, s: str) -> list[int]:
        n = len(s)
        z = [0]*n
        l=r=0
        for i in range(1, n):
            if i > r:
                l=r=i
                while r < n and s[r] == s[r-l]:
                    r+=1
                z[i] = r-l
                r-=1
            else:
                k = i-l
                if z[k] < r-i+1:
                    z[i] = z[k]
                else:
                    l = i
                    while r < n and s[r] == s[r-l]:
                        r+=1
                    z[i] = r-l
                    r-=1
        return z

    def _seralize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return "$N"
        return f"${root.val}{self._seralize(root.left)}{self._seralize(root.right)}"

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        substr = self._seralize(subRoot)
        text = self._seralize(root)
        z = self.z_function(f"{substr}|{text}")
        k = len(substr)
        for i in range(len(text)):
            if z[i+k+1] == k:
                return True
        return False

File: DSA/src/go/subtree_of_another_tree.go
package main

import "strconv"

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	var serialize func(node *TreeNode) string
	serialize = func(node *TreeNode) string {
		if node == nil {
			return "#$"
		}
		return "#" + strconv.Itoa(node.Val) + serialize(node.Left) + serialize(node.Right)
	}

	var z_func func(s string) []int
	z_func = func(s string) []int {
		n := len(s)
		z := make([]int, n)
		l, r := 0, 0
		for i := 1; i < n; i++ {
			if i > r {
				l, r = i, i
				for r < n && s[r] == s[r-l] {
					r++
				}
				z[i] = r - l
				r--
			} else {
				k := i - l
				if z[k] < r-i+1 {
					z[i] = z[k]
				} else {
					l = i
					for r < n && s[r] == s[r-l] {
						r++
					}
					z[i] = r - l
					r--
				}
			}
		}
		return z
	}

	substr, str := serialize(subRoot), serialize(root)

	z := z_func(substr + "|" + str)
	k := len(substr)
	for i := 0; i < len(str); i++ {
		if z[i+k+1] == k {
			return true
		}
	}
	return false
}
